In [10]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [11]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [12]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [13]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [14]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
crypto_scaled = StandardScaler().fit_transform(df_market_data)

crypto_scaled

array([[ 5.08529366e-01,  4.93193071e-01,  7.72200433e-01,
         2.35459633e-01, -6.74950963e-02, -3.55953481e-01,
        -2.51636882e-01],
       [ 1.85445894e-01,  9.34445040e-01,  5.58692121e-01,
        -5.43409317e-02, -2.73482725e-01, -1.15759474e-01,
        -1.99352110e-01],
       [ 2.17739616e-02, -7.06336853e-01, -2.16804207e-02,
        -6.10301536e-02,  8.00452481e-03, -5.50246924e-01,
        -2.82060506e-01],
       [-4.07643829e-02, -8.10928066e-01,  2.49457974e-01,
        -5.03879651e-02, -3.73164019e-01, -4.58258816e-01,
        -2.95546142e-01],
       [ 1.19303608e+00,  2.00095907e+00,  1.76061001e+00,
         5.45842065e-01, -2.91202870e-01, -4.99847761e-01,
        -2.70316950e-01],
       [ 8.91870708e-01,  1.32729453e+00,  8.00214184e-01,
        -5.71478992e-02,  7.78653106e-01, -1.88231917e-01,
        -2.25532605e-01],
       [ 1.13972400e-02,  2.57225091e+00,  1.10164693e+00,
        -4.90495415e-01, -9.31954023e-01,  3.87758986e-01,
        -1.8284399

In [15]:
# Create a DataFrame with the scaled data
crypto_transformed = pd.DataFrame(crypto_scaled, columns=df_market_data.columns)

# Copy the crypto names from the original data
crypto_transformed["coin_id"]= df_market_data.index

# Set the coinid column as index
crypto_transformed = crypto_transformed.set_index("coin_id")

# Display sample data
crypto_transformed.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [16]:
# Create a list with the number of k-values from 1 to 11
k=range(1,12)
list(k) 

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [17]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k: 
    model = KMeans(n_clusters=i)
    model.fit(crypto_transformed)
    inertia.append(model.inertia_)

inertia

c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting t

[287.0,
 198.57181809798126,
 123.19048183836959,
 79.02243535120975,
 63.85866780584265,
 52.625107998757855,
 44.178820652774974,
 39.112748395210446,
 32.120944114053906,
 28.34780555208513,
 24.430452921041606]

In [18]:
# Create a dictionary with the data to plot the Elbow curve
dic_inertia={'k':k, 'inertia': inertia}

# Create a DataFrame with the data to plot the Elbow curve
inertia_df=pd.DataFrame(dic_inertia)

inertia_df

k     inertia
0    1  287.000000
1    2  198.571818
2    3  123.190482
3    4   79.022435
4    5   63.858668
5    6   52.625108
6    7   44.178821
7    8   39.112748
8    9   32.120944
9   10   28.347806
10  11   24.430453

In [19]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_data1 = inertia_df.hvplot.line(
    x='k',
    y= 'inertia',
    title= 'Elbow_Curve',
    xticks= k
)    

elbow_data1

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** Based on the elbow rule, 4 would be the best value for 'k'!

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [20]:
# Initialize the K-Means model using the best value for k
model= KMeans(n_clusters=4)

In [21]:
# Fit the K-Means model using the scaled data
model.fit(crypto_transformed)

c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4)

In [22]:
# Predict the clusters to group the cryptocurrencies using the scaled data
clusters= model.predict(crypto_transformed)

# Print the resulting array of cluster values.
clusters

array([3, 3, 0, 0, 3, 3, 3, 3, 3, 0, 0, 0, 0, 3, 0, 3, 0, 0, 3, 0, 0, 3,
       0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 2, 3, 0, 0, 1, 0, 0, 0, 0])

In [23]:
# Create a copy of the DataFrame
clusters_predicted = crypto_transformed.copy()

In [24]:
# Add a new column to the DataFrame with the predicted clusters
clusters_predicted['predicted_cluster']= clusters

# Display sample data
clusters_predicted.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  predicted_cluster  
coin_id                                                      
bitcoin                        -0.251637                  3  
ethereum                       -0.199352                  3  
tether                         -0.282061                  0  
ripple                         -0.295546                  0  
bitcoin-cash                   -0.270317                  3

In [25]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
clusters_predicted.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by= "predicted_cluster", 
    hover_cols = "coin_id"
)

:NdOverlay   [predicted_cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [27]:
# Create a PCA model instance and set `n_components=3`.
pca= PCA(n_components=3)

In [28]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
clusters_pca = pca.fit_transform(crypto_transformed)

# View the first five rows of the DataFrame. 
clusters_pca[0:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [29]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 
Total Explained Variance:0.3719856+0.34700813+0.17603793=0.89503166

Total Explained Variance = 0.89503166

The total explained variance of the three principal components is approximately 0.8950 or 89.50%.

In [31]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data
clusters_pca_df=pd.DataFrame(
    clusters_pca,
    columns=["PCA1", "PCA2", "PCA3"]
)
clusters_pca_df

# Copy the crypto names from the original data
clusters_pca_df['coin_id']= list(clusters_predicted.index)
clusters_pca_df

# Set the coinid column as index
clusters_pca_df = clusters_pca_df.set_index('coin_id')

# Display sample data
clusters_pca_df

PCA1      PCA2      PCA3
coin_id                                           
bitcoin              -0.600667  0.842760  0.461595
ethereum             -0.458261  0.458466  0.952877
tether               -0.433070 -0.168126 -0.641752
ripple               -0.471835 -0.222660 -0.479053
bitcoin-cash         -1.157800  2.041209  1.859715
binancecoin          -0.516534  1.388377  0.804071
chainlink            -0.450711  0.517699  2.846143
cardano              -0.345600  0.729439  1.478013
litecoin             -0.649468  0.432165  0.600303
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653
crypto-com-chain     -0.248198 -1.376252 -1.462026
usd-coin             -0.438408 -0.175337 -0.663388
eos                  -0.693425 -0.473815 -0.527597
monero                0.060499  2.909404  1.498571
tron                 -0.393352 -0.108192 -0.012756
tezos                -0.796176 -0.494409  1.082812
okb                   0.064075 -1.269825 -1.098829
stellar              -0.489015 -0.732719 -0.062543
cosmos               -0.306272  0.703415  1.714224
cdai                 -0.513528 -0.142802 -0.656566
neo                  -0.362120 -0.986914 -0.728752
wrapped-bitcoin      -0.604265  0.827398  0.439316
leo-token            -0.413296 -0.674115 -1.076628
huobi-token          -0.407483 -0.212507 -0.351426
nem                   0.608974  0.563532 -1.148742
binance-usd          -0.450211 -0.151019 -0.647401
iota                 -0.764665 -0.517886  0.204990
vechain              -0.556315 -1.938209 -1.261776
zcash                -0.425147  0.492976  1.058048
theta-token           2.676868 -0.013954 -1.965207
dash                 -0.613923 -0.479337  0.339565
ethereum-classic     -0.579924 -0.356334 -0.114942
ethlend               8.089018 -3.896891  2.301382
maker                -0.389045  0.165041  0.379414
havven                0.865762 -2.261882  0.275583
omisego               0.111675  0.428316 -1.205398
celsius-degree-token  4.792395  6.767679 -1.986985
ontology             -0.632355 -2.108117 -0.652227
ftx-token            -0.593142  0.021485  0.209911
true-usd             -0.458131 -0.135734 -0.635284
digibyte             -0.297910 -0.191126 -0.909602

---

### Find the Best Value for k Using the PCA Data

In [32]:
# Create a list with the number of k-values from 1 to 11
k= list(range(1,12))
k

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [33]:
# Create an empty list to store the inertia values
inertia2 = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k: 
    k_model= KMeans(n_clusters=i, random_state=1)
    k_model.fit(clusters_pca_df)
    inertia2.append(k_model.inertia_)

inertia2

c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting t

[256.8740855678926,
 165.90199402036018,
 93.77462568057304,
 49.6654966517974,
 38.35225121638687,
 27.61897178795745,
 21.134056037473616,
 17.43766423702042,
 13.742791960480568,
 10.484890485976928,
 8.114985790821041]

In [34]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data2= {"k": k, "inertia": inertia2}

# Create a DataFrame with the data to plot the Elbow curve
elbow_df2= pd.DataFrame(elbow_data2)

elbow_df2

k     inertia
0    1  256.874086
1    2  165.901994
2    3   93.774626
3    4   49.665497
4    5   38.352251
5    6   27.618972
6    7   21.134056
7    8   17.437664
8    9   13.742792
9   10   10.484890
10  11    8.114986

In [35]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow2_plot= elbow_df2.hvplot.line(
    x="k",
    y="inertia",
    title= "Elbow_Curve",
    xticks= k
)

elbow2_plot

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** The best value for `k` when using the PCA data is 4. 


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, it does not differ from the best k value found when using the original data

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [36]:
# Initialize the K-Means model using the best value for k
model=  KMeans(n_clusters=4, random_state=1)

In [37]:
# Fit the K-Means model using the PCA data
model.fit(clusters_pca_df)

c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
c:\Users\Krissy\anaconda3-1\Lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [38]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_4=model.predict(clusters_pca_df)

# Print the resulting array of cluster values.
k_4

array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0])

In [39]:
# Create a copy of the DataFrame with the PCA data
copy_cluster_pca_df= clusters_pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
copy_cluster_pca_df['clusters_predicted'] = k_4

# Display sample data
copy_cluster_pca_df

PCA1      PCA2      PCA3  clusters_predicted
coin_id                                                               
bitcoin              -0.600667  0.842760  0.461595                   1
ethereum             -0.458261  0.458466  0.952877                   1
tether               -0.433070 -0.168126 -0.641752                   0
ripple               -0.471835 -0.222660 -0.479053                   0
bitcoin-cash         -1.157800  2.041209  1.859715                   1
binancecoin          -0.516534  1.388377  0.804071                   1
chainlink            -0.450711  0.517699  2.846143                   1
cardano              -0.345600  0.729439  1.478013                   1
litecoin             -0.649468  0.432165  0.600303                   1
bitcoin-cash-sv      -0.759014 -0.201200 -0.217653                   0
crypto-com-chain     -0.248198 -1.376252 -1.462026                   0
usd-coin             -0.438408 -0.175337 -0.663388                   0
eos                  -0.693425 -0.473815 -0.527597                   0
monero                0.060499  2.909404  1.498571                   1
tron                 -0.393352 -0.108192 -0.012756                   0
tezos                -0.796176 -0.494409  1.082812                   1
okb                   0.064075 -1.269825 -1.098829                   0
stellar              -0.489015 -0.732719 -0.062543                   0
cosmos               -0.306272  0.703415  1.714224                   1
cdai                 -0.513528 -0.142802 -0.656566                   0
neo                  -0.362120 -0.986914 -0.728752                   0
wrapped-bitcoin      -0.604265  0.827398  0.439316                   1
leo-token            -0.413296 -0.674115 -1.076628                   0
huobi-token          -0.407483 -0.212507 -0.351426                   0
nem                   0.608974  0.563532 -1.148742                   0
binance-usd          -0.450211 -0.151019 -0.647401                   0
iota                 -0.764665 -0.517886  0.204990                   0
vechain              -0.556315 -1.938209 -1.261776                   0
zcash                -0.425147  0.492976  1.058048                   1
theta-token           2.676868 -0.013954 -1.965207                   0
dash                 -0.613923 -0.479337  0.339565                   0
ethereum-classic     -0.579924 -0.356334 -0.114942                   0
ethlend               8.089018 -3.896891  2.301382                   2
maker                -0.389045  0.165041  0.379414                   1
havven                0.865762 -2.261882  0.275583                   0
omisego               0.111675  0.428316 -1.205398                   0
celsius-degree-token  4.792395  6.767679 -1.986985                   3
ontology             -0.632355 -2.108117 -0.652227                   0
ftx-token            -0.593142  0.021485  0.209911                   0
true-usd             -0.458131 -0.135734 -0.635284                   0
digibyte             -0.297910 -0.191126 -0.909602                   0

In [40]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
clusters_predicted2= copy_cluster_pca_df.hvplot.scatter(
    x="PCA1",
    y= "PCA2",
    by= 'clusters_predicted',
    hover_cols= 'coin_id'
)

clusters_predicted2

:NdOverlay   [clusters_predicted]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [41]:
# Composite plot to contrast the Elbow curves
inertia_df.hvplot.line(x="k", y="inertia", title= "Elbow_Curve_1", xticks=k) + elbow_df2.hvplot.line(x="k", y="inertia",title="Elbow_Curve_2", xticks=k)


:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [42]:
# Composite plot to contrast the clusters
clusters_predicted.hvplot.scatter(x= 'price_change_percentage_24h', y='price_change_percentage_7d', by='predicted_cluster', hover_cols= 'coin_id')+ copy_cluster_pca_df.hvplot.scatter(x= 'PCA1', y= 'PCA2', by= 'clusters_predicted', hover_cols= 'coin_id')

:Layout
   .NdOverlay.I  :NdOverlay   [predicted_cluster]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [clusters_predicted]
      :Scatter   [PCA1]   (PCA2,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Using fewer features to cluster the data can simplify the model and improve interpretability and readability by making clusters easier to visualize and understand. However, this reduction can lead to the loss of important data and, potentially result in less accurate clusters and more overlapping groups. 